plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/util/TestingHadoop.java

```diff
@@ -268,4 +275,19 @@ public TestingHadoop build()
             return new TestingHadoop(image, ports, resourcesToMount, filesToMount, extraHosts, network, retryLimit);
         }
     }
+
+    public static Path renderHadoopCoreSiteTemplate(String templatePath, Map<String, String> replacements)
+            throws IOException
+    {
+        String coreSiteXmlContent = Resources.toString(Resources.getResource(templatePath), UTF_8);
+        for (Map.Entry<String, String> replacement : replacements.entrySet()) {
+            coreSiteXmlContent = coreSiteXmlContent.replace(replacement.getKey(), replacement.getValue());
```

Comment:
The result may depend on the iteration order over `replacement` if a replaced value contains replacement key for some other replacement.

The previous code didn't exhibit this problem because it knew the values it operated on (the replacement keys and the potential replacement values). The new code looks more abstract, so it no longer can safely do such assumptions.

Reply:
Should we revert to the old code and include another `core-site.xml` for Ceph (e.g. `core-site-ceph.xml`?

Or we can throw an exception when the replaced value contains a replacement key for some other replacement?

Reply:
The correct way to apply replacement without inadvertently allowing recursive replacement would be

```java
String coreSiteXmlContent = Resources.toString(Resources.getResource(templatePath), UTF_8);
if (!replacements.isEmpty()) {
    StringBuilder replaced = new StringBuilder();
    Pattern keyPattern = Pattern.compile(replacements.keySet().stream()
            .map(Pattern::quote)
            .collect(joining("|")));
    Matcher matcher = keyPattern.matcher(coreSiteXmlContent);
    while (matcher.find()) {
        String replacementKey = matcher.group(0);
        String replacementValue = replacements.get(replacementKey);
        verifyNotNull(replacementValue, "No replacement value for '%s' in the map: %s", replacementKey, replacements);
        matcher.appendReplacement(replaced, replacementValue);
    }
    matcher.appendTail(replaced);
    coreSiteXmlContent = replaced.toString();
}
```